posts/loggingPython.md

```diff
@@ -0,0 +1,325 @@
+
+```python
+def search(input_filename, output_filename, text):
+    # overwrite output file
+    with open(output_filename, "w") as out_file:
+        with open(input_filename) as in_file:
+            # loop over each log line and check if text appears
+            for line in in_file:
+                if text in line:
+                    out_file.write(line)
+```
+
+This makes it easy to search for log messages that contain keywords such as `critical` or `warning`.
+
+### Rotating Logs
+
+Though not generally taught in logging guides, the Python logging module makes it easy to log in a different file after an interval of time or after the log file reaches a certain size.
+
+This becomes useful if you automatically want to get rid of older logs, or if you're going to search through your logs by date since you won’t have to search through a huge file to find a set of logs that are already grouped.
+
+To create a `handler` that makes a new log file every day and automatically deletes logs more than five days old, you can use a `TimedRotatingFileHandler`. Here’s an example:
+
+```python
+logger = logging.getLogger('Rotating Log by Day')
+
+# writes to pathToLog
+# creates a new file every day because `when="d"` and `interval=1`
+# automatically deletes logs more than 5 days old because `backupCount=5`
+handler = TimedRotatingFileHandler(pathToLog, when="d", interval=1,  backupCount=5)
+```
+
+### Accessing Server Logs through SSH
+
+When troubleshooting an issue on your server, you might have to check your error logs. The best way to do that is to SSH into your server.
+
+If you don’t know how to SSH into your server, [here](https://help.dreamhost.com/hc/en-us/articles/216041267-SSH-overview) is an excellent guide that explains how.
+
+Once you are connected, you can go to the `logs` folder.
+```bash
+cd logs
+```
+
+ You can use `ls` to list all the types of logs you can view and `cd` into the relevant domain.
+
+To see the last 20 logs from `error.log`:
+```bash
+tail error.log
+```
+
+To show new logs in real-time:
+```bash
+tail -f error.log
+```
+
+To search for a specific keyword in a log file, you can use the `grep` command.
+
+To see all logs that have `warning` in `error.log`:
+```bash
+cat error.log | grep "warning"
+```
+
+## Wrapping Up
+
+Logs can be a pain to deal with. That’s why we recommend a [cloud-based provider](https://timber.io/) that can deal with aggregating the logs and making them intuitive, but it’s difficult to try new technology when the payoff isn’t visible.
+
+### Logs as a Source of Truth
+
+If you take anything away from this post, it should be that logs serve as the _source of truth_ for the user’s actions. Even for ephemeral actions, such as putting an item into and out of a cart, it’s essential to be able to trace the user’s steps during a bug report, and the logs allow you to determine their actions between all your [MicroServices](https://timber.io/blog/docker-and-the-rise-of-microservices/).
```

Comment:
This comment should be hoisted to the beginning of the article. I would also prefer it if you remove the reference to microservices.